src/ptw.cc

```diff
@@ -0,0 +1,507 @@
+
+			assert(CR3_addr != UINT64_MAX);
+			PageTablePage* curr_page = L5; //Start wth the L5 page
+			uint64_t next_level_base_addr = UINT64_MAX;
+			bool page_fault = false;
+
+			for (int i = 5; i > MSHR.entry[index].translation_level; i--)
+			{
+				uint64_t offset = get_offset(MSHR.entry[index].full_v_addr, i); //Get offset according to page table level
+				assert(curr_page != NULL);
+				next_level_base_addr = curr_page->next_level_base_addr[offset];
+				if(next_level_base_addr == UINT64_MAX)
+				{
+					handle_page_fault(curr_page, &MSHR.entry[index], i); //i means next level does not exist.
+					page_fault = true;
+					MSHR.entry[index].translation_level = 0; //In page fault, All levels are translated.
+					break;
+				}
+				curr_page = curr_page->entry[offset];
+			}
+
+			if(MSHR.entry[index].translation_level == 0) //If translation complete
+			{
+				curr_page = L5;
+				next_level_base_addr = UINT64_MAX;
+				for (int i = 5; i > 1; i--) //Walk the page table and fill MMU caches
+				{
+					uint64_t offset = get_offset(MSHR.entry[index].full_v_addr, i);
+					assert(curr_page != NULL);
+					next_level_base_addr = curr_page->next_level_base_addr[offset];
+					assert(next_level_base_addr != UINT64_MAX);
+					curr_page = curr_page->entry[offset];
+
+					if(MSHR.entry[index].init_translation_level - i >= 0) //Check which translation levels needs to filled
+					{
+						switch(i)
+						{
+							case 5: fill_mmu_cache(PSCL5, next_level_base_addr, &MSHR.entry[index], IS_PSCL5);
+									break;
+							case 4: fill_mmu_cache(PSCL4, next_level_base_addr, &MSHR.entry[index], IS_PSCL4);
+									break;
+							case 3: fill_mmu_cache(PSCL3, next_level_base_addr, &MSHR.entry[index], IS_PSCL3);
+									break;
+							case 2: fill_mmu_cache(PSCL2, next_level_base_addr, &MSHR.entry[index], IS_PSCL2);
+									break;
+						}
+					}
+				}
+
+				uint64_t offset = get_offset(MSHR.entry[index].full_v_addr, IS_PTL1);
+				next_level_base_addr = curr_page->next_level_base_addr[offset];
+
+				MSHR.entry[index].event_cycle = current_core_cycle[cpu]; //Page fault are completed in same cycle	
+
+
+				MSHR.entry[index].data = next_level_base_addr << LOG2_PAGE_SIZE | (MSHR.entry[index].full_v_addr & ((1<<LOG2_PAGE_SIZE) - 1)); //Return the translated physical address to STLB
+			
+				for(auto ret: MSHR.entry[index].to_return)
+					ret->return_data(&MSHR.entry[index]);
+
+				if(warmup_complete[cpu])
+			      {
+					uint64_t current_miss_latency = (current_core_cycle[cpu] - MSHR.entry[index].cycle_enqueued);	
+					total_miss_latency += current_miss_latency;
+			      }
+
+				MSHR.remove_queue(&MSHR.entry[index]);
+			}
+			else
+			{
+				assert(!page_fault); //If page fault was there, then all levels of translation should have be done.
+
+				if((((CACHE*)lower_level)->RQ.occupancy() < ((CACHE*)lower_level)->RQ.size())) //Lower level of PTW is L2C. If L2 RQ has space then send the next level of translation.
```

Comment:
Rather than this if statement, you should rely on the underlying `MemoryRequestConsumer::add_rq()` to return -2 if the queue is full. This permits entries to be merged, which can be particularly important for high-order page table pages.